packages/vm/src/imports.rs

```diff
@@ -215,41 +176,29 @@ pub fn do_scan<S: Storage + 'static, Q: Querier>(
         .try_into()
         .map_err(|_| CommunicationError::invalid_order(order))?;
 
-    let ffi_result = with_storage_from_context::<S, Q, _, _>(ctx, |store| {
+    let (result, gas_info) = with_storage_from_context::<S, Q, _, _>(ctx, |store| {
         Ok(store.range(start.as_deref(), end.as_deref(), order))
     })?;
-    let iterator_id = match ffi_result {
-        Ok((iterator, gas_info)) => {
-            process_gas_info::<S, Q>(ctx, gas_info)?;
+    process_gas_info::<S, Q>(ctx, gas_info)?;
+    let iterator_id = match result {
+        Ok(iterator) => {
             let new_id = add_iterator::<S, Q>(ctx, iterator);
             Ok(new_id)
         }
-        Err((err, gas_info)) => {
-            process_gas_info::<S, Q>(ctx, gas_info)?;
-            Err(VmError::from(err))
-        }
+        Err(err) => Err(VmError::from(err)),
     }?;
 
     Ok(iterator_id)
 }
 
 #[cfg(feature = "iterator")]
 pub fn do_next<S: Storage, Q: Querier>(ctx: &mut Ctx, iterator_id: u32) -> VmResult<u32> {
-    let ffi_result =
+    let (result, gas_info) =
         with_iterator_from_context::<S, Q, _, _>(ctx, iterator_id, |iter| Ok(iter.next()))?;
-    let kv = match ffi_result {
-        Ok((kv, gas_info)) => {
-            process_gas_info::<S, Q>(ctx, gas_info)?;
-            Ok(kv)
-        }
-        Err((err, gas_info)) => {
-            process_gas_info::<S, Q>(ctx, gas_info)?;
-            Err(VmError::from(err))
-        }
-    }?;
+    process_gas_info::<S, Q>(ctx, gas_info)?;
 
     // Empty key will later be treated as _no more element_.
-    let (key, value) = kv.unwrap_or_else(|| (Vec::<u8>::new(), Vec::<u8>::new()));
+    let (key, value) = (result?).unwrap_or_else(|| (Vec::<u8>::new(), Vec::<u8>::new()));
```

Comment:
You don't need the `()` around `result?`

Reply:
Done in https://github.com/CosmWasm/cosmwasm/pull/489/commits/922f9f5b8a065cb9d3d8b86d1d0318b2b6f43d27